chips/stm32f4xx/src/i2c.rs

```diff
@@ -0,0 +1,462 @@
+    I2C1_BASE,
+    I2CClock(rcc::PeripheralClock::APB1(rcc::PCLK1::I2C1)),
+);
+
+impl I2C<'_> {
+    const fn new(base_addr: StaticRef<I2CRegisters>, clock: I2CClock) -> Self {
+        Self {
+            registers: base_addr,
+            clock,
+
+            master_client: OptionalCell::empty(),
+
+            slave_address: Cell::new(0),
+
+            buffer: TakeCell::empty(),
+            tx_position: Cell::new(0),
+            rx_position: Cell::new(0),
+
+            tx_len: Cell::new(0),
+            rx_len: Cell::new(0),
+
+            status: Cell::new(I2CStatus::Idle),
+        }
+    }
+
+    pub fn set_speed(&self, speed: I2CSpeed, system_clock_in_mhz: usize) {
+        self.disable();
+        self.registers
+            .cr2
+            .modify(CR2::FREQ.val(system_clock_in_mhz as u32));
+        match speed {
+            I2CSpeed::Speed100k => {
+                self.registers
+                    .ccr
+                    .modify(CCR::CCR.val(system_clock_in_mhz as u32 * 5) + CCR::FS::SM_MODE);
+                self.registers
+                    .trise
+                    .modify(TRISE::TRISE.val(system_clock_in_mhz as u32 + 1));
+            }
+            I2CSpeed::Speed400k => {
+                self.registers
+                    .ccr
+                    .modify(CCR::CCR.val(system_clock_in_mhz as u32 * 5 / 6) + CCR::FS::FM_MODE);
+                self.registers
+                    .trise
+                    .modify(TRISE::TRISE.val(system_clock_in_mhz as u32 + 1));
+            }
+        }
+        self.enable();
+    }
+
+    pub fn is_enabled_clock(&self) -> bool {
+        self.clock.is_enabled()
+    }
+
+    pub fn enable_clock(&self) {
+        self.clock.enable();
+    }
+
+    pub fn disable_clock(&self) {
+        self.clock.disable();
+    }
+
+    pub fn handle_event(&self) {
+        if self.registers.sr1.is_set(SR1::SB) {
+            let dir = match self.status.get() {
+                I2CStatus::Writing | I2CStatus::WritingReading => 0,
+                I2CStatus::Reading => 1,
+                _ => panic!("invalid i2c state when setting address"),
+            };
+            self.registers
+                .dr
+                .write(DR::DR.val(((self.slave_address.get() << 1) as u32) | dir));
+        }
+        if self.registers.sr1.is_set(SR1::ADDR) {
+            // i2c requires a sr2 read
+            self.registers.sr2.get();
+        }
+        if self.registers.sr1.is_set(SR1::TXE) {
+            // send the next byte
+            if self.buffer.is_some() && self.tx_position.get() < self.tx_len.get() {
+                self.buffer.map(|buf| {
+                    let byte = buf[self.tx_position.get() as usize];
+                    self.registers.dr.write(DR::DR.val(byte as u32));
+                    self.tx_position.set(self.tx_position.get() + 1);
+                });
+            }
+        }
+
+        while self.registers.sr1.is_set(SR1::RXNE) {
```

Comment:
I'm a bit confused by this loop -- does this effectively become a synchronous transfer at this point?

Reply:
As far as I understand, the i2c implemented in stm32f4xx has a byte in the data register and might immediatly have another one in the shift register. The idea was to process both bytes without needing another interrupt cycle.

Reply:
That makes sense -- in theory then, if in some universe the I2C bus were fast enough and the core were slow enough, this would become synchronous -- but without this it would keep throwing interrupts for new data anyway.. okay, that all makes sense to me. Thanks!